src/compiler/checker.ts

```diff
@@ -9801,8 +9888,13 @@ namespace ts {
             //
             // For a decorator, no arguments are susceptible to contextual typing due to the fact
             // decorators are applied to a declaration by the emitter, and not to an expression.
+            let excludeCallee: boolean;
             let excludeArgument: boolean[];
             if (!isDecorator) {
+                const calleeNode = node.kind === SyntaxKind.CallExpression && (<PropertyAccessExpression>(<CallExpression>node).expression).expression;
+                if (calleeNode && isContextSensitive(calleeNode)) {
+                    excludeCallee = true;
+                }

```

Comment:
Expressions like  `(function (n) { return this.m + n }).call({ m: 101 }, 12)` should contextually give a type to `this.m`. But I guess this won't work if the function's type is already determined. I should look at my tests again, I thought I had one that showed this in action.